### PR TITLE
Feature/intolerances

### LINKED ISSUE #6: User Story 5: Adding intolerances
## Value statement 
As a **user** 
I need **the possibility to save intolerances etc. from my guests** 
so that **I don't forget them and to automatically set the recipe-filters later on**. 

## Description
CreateGuest.js (#3 ):
image

EditGuest.js (#5 ):
image

GuestDetails.js (#4 ):
image

## Acceptance criteria
- [x] it is possible to add intolerances on the EditGuest- and CreateGuest-pages 
- [x] possible intolerances are shown beneath the input-field
- [x] clicking on proposed intolerances adds them to an "active"-list
- [x] clicking on submit saves the intolerances for this specific guest
- [x] you see the list of intolerances on the GuestDetails-page

## Tasks
- [x] create an AddIntolerances-Component with an empty list, an input-field and an empty div for search results
- [x] create an intolerances-array with all 34 available intolerances
- [x] npm install fuzzy-search
- [x] filter the intolerances according to the user-input and render them beneath the input field (max. 3)
- [x] add proposed intolerances to an "active"-list onClick
- [x] save the intolerances for this specific guest onSubmit
- [x] render the list of intolerances on the GuestDetails-page 
- [x] implement the AddIntolerances-Component in CreateGuest.js and EditGuest.js

## Size
big

### PR DESCRIPTION
**Pull request by s-kond**
fixes #6

### CODE REVIEW

File: src/pages/CreateGuest.js [hunk 1]
```diff
@@ -12,14 +20,47 @@ export default function CreateGuest({ onHandleSubmit }) {
...
       alert("Each name must have one letter at least.");
     } else {
-      onHandleSubmit(newName.value, newNotes.value);
+      onHandleSubmit(newName.value, newNotes.value, activeList);
       navigate("/");
     }
   }
 
+  function compareArrays(userIntolerances, allPossibleIntolerances) {
+    let editedPossibleIntolerances = [];
```
Comment (mpagels) [inline src/pages/CreateGuest.js]: - Do you net `let` here
- why do you call the variable `editied...` when it is seems to just a `filtered` array. You did not edit somehting here. But I might be wrong.

Anyway, I challenge you to try to write this function as an one-liner. Might wanna check [array filter methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter?retiredLocale=de)

This function could be exported and tested. ;-) ❤️

File: src/pages/CreateGuest.js [hunk 2]
```diff
@@ -29,6 +70,37 @@ export default function CreateGuest({ onHandleSubmit }) {
...
+          id="newIntolerances"
+          type="text"
+          placeholder="e.g. fish-free"
+          ref={intolerancesRef}
+          onChange={(event) => searchIntolerance(event.target.value)}
+        />
+        <section>
+          {activeList.map((item) => (
+            <StyledActiveIntolerance
+              key={nanoid()}
```
Comment (mpagels) [inline src/pages/CreateGuest.js]: using `nanoid` here is not a good idea. React uses these ids, to make intern some comparing, to know, how to optimal update the DOM. When, after every rendering the id changes, this could be hard for react and is very like to create some errors or unwanted behavior. I highly recommend using unique ids, that are not chaning on every render.

File: src/pages/CreateGuest.js [hunk 3]
```diff
@@ -29,6 +70,37 @@ export default function CreateGuest({ onHandleSubmit }) {
...
+              onClick={() => removeFromActive(item)}
+            >
+              {item}
+            </StyledActiveIntolerance>
+          ))}
+        </section>
+        <section>
+          {filteredIntolerance.map((item) => (
+            <StyledSearchResult
+              key={nanoid()}
```
Comment (mpagels) [inline src/pages/CreateGuest.js]: here as well. see my comment above.

File: src/pages/EditGuest.js [hunk 4]
```diff
@@ -19,11 +30,49 @@ export default function EditGuest({ guestData, onHandleEditSubmit }) {
...
+    let editedPossibleIntolerances = [];
+    for (let i = 0; i < allPossibleIntolerances.length; i++) {
+      if (userIntolerances.includes(allPossibleIntolerances[i])) {
+        continue;
+      } else {
+        editedPossibleIntolerances.push(allPossibleIntolerances[i]);
+      }
+    }
+    setEditedIntolerances(editedPossibleIntolerances);
+  }
```
Comment (mpagels) [inline src/pages/EditGuest.js]: Since I saw this code in another component, you want to DRY. So export this `utiliy` function and import it, when you needed it. This has the advantage, if you change somehting in the logic, you have to change it only in one place.

Testing the function is then easy as well.

File: src/pages/EditGuest.js [hunk 5]
```diff
@@ -40,24 +89,41 @@ export default function EditGuest({ guestData, onHandleEditSubmit }) {
...
+          name="newIntolerances"
+          id="newIntolerances"
+          type="text"
+          ref={intolerancesRef}
+          onChange={(event) => searchIntolerance(event.target.value)}
+        />
+        <section>
+          {activeEditList.map((item) => (
+            <StyledActiveIntolerance
+              key={nanoid()}
```
Comment (mpagels) [inline src/pages/EditGuest.js]: `nanoid`

File: src/pages/EditGuest.js [hunk 6]
```diff
@@ -40,24 +89,41 @@ export default function EditGuest({ guestData, onHandleEditSubmit }) {
...
+              onClick={() => removeFromActive(item)}
+            >
+              {item}
+            </StyledActiveIntolerance>
+          ))}
+        </section>
+        <section>
+          {filteredIntolerance.map((item) => (
+            <StyledSearchResult
+              key={nanoid()}
```
Comment (mpagels) [inline src/pages/EditGuest.js]: `nanoid`

File: src/pages/GuestDetails.js [hunk 7]
```diff
@@ -21,6 +22,12 @@ export default function GuestDetails({ onDelete }) {
           Edit
         </StyledEditButton>
         <StyledName>{name}</StyledName>
+        <h3>Food should be:</h3>
+        <StyledList>
+          {intolerances.map((item) => (
+            <li key={nanoid()}>{item}</li>
```
Comment (mpagels) [inline src/pages/GuestDetails.js]: `nanoid`

**Review by mpagels — CHANGES_REQUESTED** [review]
Hey Stefan, very awesome work here. A lot of interesting logic here.

I found some issues regardind `nanoid` as a key prop. I think this is not a good idea. See more in the comments. I left a challenge as well ;-) And some small improvements. Pls review.

File: src/pages/CreateGuest.js (hunk 1 above)
Comment (s-kond) [reply src/pages/CreateGuest.js]: I try 👍

File: src/util/CompareArrays.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+//a function to compare two arrays with objects ->
+//array1 (userIntolerances) containing objects that should be deleted from array2 (allPossibleIntolerances)
```
Comment (mpagels) [inline src/util/CompareArrays.js]: I really like that you want to explain what the function is doing. Try this `best practise` format instead:

```js
/**
 * Returns a filtered array.
 *
 * @param {array} userIntolerances An array with user intorlerances.
 * @param {array} allPossibleIntolerances Another array with infos in it.
 * @return {array}  a filtered array.
 */

...code
```

You have to put in real information, I was just tinkering around.

Comment (mpagels) [reply src/util/CompareArrays.js]: aaaaaaaaaaaaaaaaaaaaand good solution. Is it working? can you write tests and check some edge cases?

**Review by mpagels — APPROVED** [review]
Very good improvement here, Stefan!

File: src/util/CompareArrays.js (hunk 8 above)
Comment (s-kond) [reply src/util/CompareArrays.js]: ok thanks :) 
I tested it manually and it works - but I'll add a test too 👍
